document/src/vespa/document/update/documentupdate.cpp

```diff
@@ -165,120 +206,179 @@ DocumentUpdate::serializeFlags(int size_) const
+    return update;
+}
+
+DocumentUpdate::UP
+DocumentUpdate::createHEAD(const DocumentTypeRepo& repo, vespalib::nbostream stream)
+{
+    auto update = std::make_unique<DocumentUpdate>();
+    update->initHEAD(repo, std::move(stream));
+    return update;
 }
 
 void
-DocumentUpdate::deserialize42(const DocumentTypeRepo& repo, ByteBuffer& buffer)
+DocumentUpdate::init42(const DocumentTypeRepo & repo, vespalib::nbostream & stream)
 {
-    int pos = buffer.getPos();
-    try{
-        buffer.getShortNetwork(_version);
+    _repo = &repo;
+    deserialize42(repo, stream);
+    reserialize();
+}
+void
+DocumentUpdate::initHEAD(const DocumentTypeRepo & repo, vespalib::nbostream && stream)
+{
+    _repo = &repo;
+    _backing = std::move(stream);
+    size_t startPos = _backing.rp();
+    deserializeHeader(repo, _backing);
+    _backing.rp(startPos);
+}
+
+void
+DocumentUpdate::initHEAD(const DocumentTypeRepo & repo, vespalib::nbostream & stream)
+{
+    size_t startPos = stream.rp();
+    deserializeHEAD(repo, stream);
+    size_t sz = stream.rp() - startPos;
+    _backing = nbostream(stream.peek() - sz, sz);
+}
 
-        std::pair<const DocumentType *, DocumentId> typeAndId(deserializeTypeAndId(repo, buffer));
+void
+DocumentUpdate::deserialize42(const DocumentTypeRepo& repo, vespalib::nbostream & stream)
+{
+    size_t pos = stream.rp();
+    try{
+        int16_t version(0);
+        stream >> version;
+        std::pair<const DocumentType *, DocumentId> typeAndId(deserializeTypeAndId(repo, stream));
         _type = typeAndId.first;
         _documentId = typeAndId.second;
         // Read field updates, if any.
-        if(buffer.getRemaining() > 0) {
+        if (! stream.empty()) {
             int sizeAndFlags = 0;
-            buffer.getIntNetwork(sizeAndFlags);
+            stream >> sizeAndFlags;
             int numUpdates = deserializeFlags(sizeAndFlags);
             _updates.reserve(numUpdates);
+            ByteBuffer buffer(stream.peek(), stream.size());
             for (int i = 0; i < numUpdates; i++) {
                 _updates.emplace_back(repo, *typeAndId.first, buffer, _version);
             }
+            stream.adjustReadPos(buffer.getPos());
         }
     } catch (const DeserializeException &) {
-        buffer.setPos(pos);
+        stream.rp(pos);
         throw;
     } catch (const BufferOutOfBoundsException &) {
-        buffer.setPos(pos);
+        stream.rp(pos);
         throw;
     }
 }
 
 void
-DocumentUpdate::deserializeHEAD(const DocumentTypeRepo &repo, ByteBuffer &buffer)
+DocumentUpdate::deserializeHeader(const DocumentTypeRepo &repo, vespalib::nbostream & stream)
+{
+    assert(_updates.empty());
+    assert(_fieldPathUpdates.empty());
+    _documentId = DocumentId(stream);
+
+    vespalib::stringref typestr = stream.peek();
+    stream.adjustReadPos(typestr.length() + 1);
+    int16_t version = 0;
+    stream >> version;
+    const DocumentType *docType = repo.getDocumentType(typestr);
+    _type = docType;
```

Comment:
`_type` will be `nullptr` at this point if the document type is not known. Throw `DocumentTypeNotFoundException` if so? Will be thrown once `deserializeXX` is called, but the early bird catches the bad document type name.

Reply:
Fixed